src/backend/distributed/metadata/metadata_sync.c

```diff
@@ -2612,3 +2612,63 @@ citus_internal_delete_shard_metadata(PG_FUNCTION_ARGS)
 
 	PG_RETURN_VOID();
 }
+
+
+/*
+ * citus_internal_update_relation_colocation is an internal UDF to
+ * delete a row in pg_dist_shard and corresponding placement rows
+ * from pg_dist_shard_placement.
+ */
+Datum
+citus_internal_update_relation_colocation(PG_FUNCTION_ARGS)
+{
+	Oid relationId = PG_GETARG_OID(0);
+	uint32 tagetColocationId = PG_GETARG_UINT32(1);
+
+	EnsureTableOwner(relationId);
+
+	if (!ShouldSkipMetadataChecks())
+	{
+		/* this UDF is not allowed allowed for executing as a separate command */
+		EnsureCoordinatorInitiatedOperation();
+
+		/* ensure that the table is in pg_dist_partition */
+		char partitionMethod = PartitionMethodViaCatalog(relationId);
+		if (partitionMethod == DISTRIBUTE_BY_INVALID)
+		{
+			/* connection from the coordinator operating on a shard */
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("The relation \"%s\" does not have a valid "
+								   "entry in pg_dist_partition.",
+								   get_rel_name(relationId))));
+		}
+		else if (partitionMethod != DISTRIBUTE_BY_HASH)
+		{
+			/* connection from the coordinator operating on a shard */
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("Updating colocation ids are only allowed for hash "
+								   "distributed tables: %c", partitionMethod)));
+		}
+
+		int count = 1;
+		List *targetColocatedTableList =
+			ColocationGroupTableList(tagetColocationId, count);
+		if (list_length(targetColocatedTableList) == 0)
+		{
+			ereport(ERROR, (errcode(ERRCODE_OBJECT_NOT_IN_PREREQUISITE_STATE),
+							errmsg("No table belongs to target colocation id: %d",
+								   tagetColocationId)));
+		}
+
+		Oid targetRelationId = linitial_oid(targetColocatedTableList);
+
+		ErrorIfShardPlacementsNotColocated(relationId, targetRelationId);
```

Comment:
I'm not sure if this check is really needed. If a malicious user wants to have a table that is marked colocated, but isn't actually colocated. Then they can easily do so by creating a new table and not matching up the shards.

Reply:
That's true, but why not prevent it as much as possible? This function would be used very infrequently, and I think we should be OK with the overhead of this check.